### PR TITLE
Serverless implementation refactor and add api-name option for serverless deployment

### PR DESCRIPTION
**Pull request by yubozhao**

### CODE REVIEW

File: bentoml/deployment/serverless/aws_lambda.py [hunk 1]
```diff
@@ -145,18 +145,23 @@ def apply(self, deployment_pb, repo, prev_deployment=None):
         if bento_python_version < minimum_python_version:
             template = 'aws-python'
 
+        if aws_config.api_name:
+            apis = [{'name': aws_config.api_name}]
```
Comment (parano) [inline bentoml/deployment/serverless/aws_lambda.py]: verify api_name in bento_config['apis']?

File: bentoml/deployment/serverless/aws_lambda.py [hunk 2]
```diff
@@ -106,6 +108,7 @@ def generate_serverless_configuration_for_aws_lambda(
             "useDownloadCache": True,
             "useStaticCache": True,
             "dockerizePip": True,
+            "layer": True,
```
Comment (parano) [inline bentoml/deployment/serverless/aws_lambda.py]: is this ok to turn on? I thought we need more customization on the lambda layer

Comment (yubozhao) [reply bentoml/deployment/serverless/aws_lambda.py]: I will remove it along with the validation check for api name

File: bentoml/deployment/serverless/aws_lambda.py [hunk 3]
```diff
@@ -145,18 +153,31 @@ def apply(self, deployment_pb, repo, prev_deployment=None):
...
+        if aws_config.api_name:
+            try:
+                ensure_api_exists_in_bento_archive_api_lists(
+                    bento_config['apis'],
+                    aws_config.api_name,
+                    deployment_spec.bento_name,
+                )
+            except BentoMLDeploymentException as error:
+                return ApplyDeploymentResponse(
+                    status=Status.INVALID_ARGUMENT(str(error))
```
Comment (parano) [inline bentoml/deployment/serverless/aws_lambda.py]: `ExceptionToReturnStatus(error) => Status`

File: bentoml/deployment/serverless/serverless_utils.py [hunk 4]
```diff
@@ -62,10 +64,18 @@ def check_nodejs_comptaible_version():
         )
 
 
+# We are using serverless framework for deployment, instead of using user's own
```
Comment (parano) [inline bentoml/deployment/serverless/serverless_utils.py]: Use python docstring:
```
def install_serverless_package():
    """
    Install serverless npm package to BentoML home directory
    """
```

File: bentoml/deployment/serverless/aws_lambda.py [hunk 5]
```diff
@@ -125,119 +130,156 @@ def generate_handler_py(bento_name, apis, output_path):
...
             )
-            logger.info(
-                'Installing additional packages: serverless-python-requirements, '
-                'serverless-apigw-binary'
+            bento_config = load_bento_service_metadata(bento_path)
+
+            template = 'aws-python3'
+            minimum_python_version = version.parse('3.0.0')
+            bento_python_version = version.parse(bento_config.env.python_version)
+            if bento_python_version < minimum_python_version:
```
Comment (parano) [inline bentoml/deployment/serverless/aws_lambda.py]: use ``` if bento_python_version < version.parse('3.0.0')```  for readability

File: bentoml/deployment/serverless/gcp_function.py [hunk 6]
```diff
@@ -61,8 +63,8 @@ def {api_name}(request):
 def generate_gcp_handler_functions_config(apis):
```
Comment (parano) [inline bentoml/deployment/serverless/gcp_function.py]: (api_names)

File: bentoml/deployment/serverless/gcp_function.py [hunk 7]
```diff
@@ -96,100 +98,126 @@ def generate_main_py(bento_name, apis, output_path):
...
+                deployment_spec.bento_name, deployment_spec.bento_version
             )
-            call_serverless_command(["deploy"], output_path)
 
-        res_deployment_pb = Deployment(state=DeploymentState())
-        res_deployment_pb.CopyFrom(deployment_pb)
-        state = self.describe(res_deployment_pb, repo).state
-        res_deployment_pb.state.CopyFrom(state)
+            bento_config = load_bento_service_metadata(bento_path)
+            apis = generate_api_list(
```
Comment (parano) [inline bentoml/deployment/serverless/gcp_function.py]:
```
apis = [gcp_config.api_name] if gcp_config.api_name is not None else bento_config.apis
```

File: bentoml/deployment/serverless/gcp_function.py [hunk 8]
```diff
@@ -96,100 +98,126 @@ def generate_main_py(bento_name, apis, output_path):
...
+                bento_config.apis, gcp_config.api_name, deployment_spec.bento_name
+            )
+            with TemporaryServerlessContent(
```
Comment (parano) [inline bentoml/deployment/serverless/gcp_function.py]: refactor this as discussed offline

File: bentoml/deployment/serverless/aws_lambda.py [hunk 9]
```diff
@@ -73,9 +78,9 @@ def {api_name}(event, context):
 def generate_aws_handler_functions_config(apis):
```
Comment (parano) [inline bentoml/deployment/serverless/aws_lambda.py]: api_names

File: bentoml/deployment/serverless/aws_lambda.py [hunk 10]
```diff
@@ -125,119 +130,156 @@ def generate_handler_py(bento_name, apis, output_path):
...
         f.write(AWS_HANDLER_PY_TEMPLATE_HEADER.format(class_name=bento_name))
         for api in apis:
-            api_content = AWS_FUNCTION_TEMPLATE.format(api_name=api['name'])
+            api_content = AWS_FUNCTION_TEMPLATE.format(api_name=api.name)
             f.write(api_content)
     return
 
 
 class AwsLambdaDeploymentOperator(DeploymentOperatorBase):
     def apply(self, deployment_pb, repo, prev_deployment=None):
```
Comment (parano) [inline bentoml/deployment/serverless/aws_lambda.py]: discussed offline, change the signature of deployment operator from `def apply(self, deployment_pb, repo, prev_deployment=None)` to `def apply(self, deployment_pb, yatai_service, prev_deployment=None)`

File: bentoml/deployment/serverless/aws_lambda.py [hunk 11]
```diff
@@ -125,119 +130,156 @@ def generate_handler_py(bento_name, apis, output_path):
...
-                'serverless-apigw-binary'
+            bento_config = load_bento_service_metadata(bento_path)
+
+            template = 'aws-python3'
+            minimum_python_version = version.parse('3.0.0')
+            bento_python_version = version.parse(bento_config.env.python_version)
+            if bento_python_version < minimum_python_version:
+                template = 'aws-python'
+
+            apis = generate_api_list(
```
Comment (parano) [inline bentoml/deployment/serverless/aws_lambda.py]: api_names = [aws_config.api_name] if aws_config.api_name is not None else bento_config.apis
verify_api_names(api_name, config)

File: bentoml/deployment/utils.py [hunk 12]
```diff
@@ -103,3 +107,24 @@ def ensure_docker_available_or_raise():
             'Docker is required for this deployment. Please visit '
             'www.docker.come for instructions'
         )
+
+
+def ensure_api_exists_in_bento_archive_api_lists(apis, api_name, bento_name):
```
Comment (parano) [inline bentoml/deployment/utils.py]: remove this function now?  or change it to verify list of apis like we discussed

File: bentoml/repository/metadata_store.py [hunk 13]
```diff
@@ -80,7 +80,7 @@ def _bento_orm_obj_to_pb(bento_obj):
         bento_obj.bento_service_metadata, BentoServiceMetadata()
     )
     bento_uri = BentoUri(
-        uri=bento_obj.uri, uri_type=BentoUri.StorageType.Value(bento_obj.uri_type)
+        uri=bento_obj.uri, type=BentoUri.StorageType.Value(bento_obj.uri_type)
```
Comment (parano) [inline bentoml/repository/metadata_store.py]: 👏

File: bentoml/deployment/serverless/gcp_function.py [hunk 14]
```diff
@@ -61,135 +62,187 @@ def {api_name}(request):
...
+            for api_name in api_names
+        },
+    }
 
     yaml.dump(serverless_config, Path(config_path))
-    return
 
 
-def generate_main_py(bento_name, apis, output_path):
+def generate_main_py(bento_name, api_names, output_path):
```
Comment (parano) [inline bentoml/deployment/serverless/gcp_function.py]: rename to `generate_gcp_function_main_py` similar to the aws code path?

File: bentoml/deployment/utils.py (hunk 12 above)
Comment (yubozhao) [reply bentoml/deployment/utils.py]: We can’t remove it for now, since we also using it in sagemaker. Let’s revisit this when we do sagemaker refactor

**Pull request merged**